### PR TITLE
[SYSTEMDS-2550] Batch scaling and weighing of imbalanced workers

### PR DESCRIPTION
**Pull request by tobiasrieger**
This PR provides a number of updates to the Federated Parameter Server

- Batch Scaling to fix imbalance during runtime
- Weighing of gradients relative to the data size the represent for all runtime balancing schemes
- Seeding for data partitioners
- New test cases and better logging

### CODE REVIEW

File: src/main/java/org/apache/sysds/runtime/instructions/cp/ParamservBuiltinCPInstruction.java [hunk 1]
```diff
@@ -143,7 +154,8 @@ private void runFederated(ExecutionContext ec) {
 		int numBatchesPerEpoch = 0;
 		if(runtimeBalancing == PSRuntimeBalancing.RUN_MIN) {
 			numBatchesPerEpoch = (int) Math.ceil(result._balanceMetrics._minRows / (float) getBatchSize());
-		} else if (runtimeBalancing == PSRuntimeBalancing.CYCLE_AVG) {
+		} else if (runtimeBalancing == PSRuntimeBalancing.CYCLE_AVG
+				|| runtimeBalancing == PSRuntimeBalancing.SCALE_BATCH) {
 			numBatchesPerEpoch = (int) Math.ceil(result._balanceMetrics._avgRows / (float) getBatchSize());
  		} else if (runtimeBalancing == PSRuntimeBalancing.CYCLE_MAX) {
 			numBatchesPerEpoch = (int) Math.ceil(result._balanceMetrics._maxRows / (float) getBatchSize());
```
Comment (sebwrede) [inline src/main/java/org/apache/sysds/runtime/instructions/cp/ParamservBuiltinCPInstruction.java:161]: To make this method shorter and easier to read, these lines could be a separate method that returns numBatchesPerEpoch.

File: src/main/java/org/apache/sysds/runtime/instructions/cp/ParamservBuiltinCPInstruction.java [hunk 2]
```diff
@@ -125,15 +127,24 @@ private void runFederated(ExecutionContext ec) {
...
+		LOG.info("[+] Runtime Balancing: " + runtimeBalancing);
+		LOG.info("[+] Weighing: " + weighing);
+		LOG.info("[+] Seed: " + seed);
 
 		// partition federated data
-		DataPartitionFederatedScheme.Result result = new FederatedDataPartitioner(federatedPSScheme)
+		DataPartitionFederatedScheme.Result result = new FederatedDataPartitioner(federatedPSScheme, seed)
 				.doPartitioning(ec.getMatrixObject(getParam(PS_FEATURES)), ec.getMatrixObject(getParam(PS_LABELS)));
 		List<MatrixObject> pFeatures = result._pFeatures;
 		List<MatrixObject> pLabels = result._pLabels;
```
Comment (sebwrede) [inline src/main/java/org/apache/sysds/runtime/instructions/cp/ParamservBuiltinCPInstruction.java:150]: Do you need to put pFeatures and pLabels in variables? It looks like it is only used once and much later in the method. The method could be shorter if you take pFeatures and pLabels directly from the result object.

**Review by sebwrede — COMMENTED** [review]
It generally looks good, I added a few comments. 

It could make sense to add documentation to each of the subclasses of DataPartitionFederatedScheme describing the partitioning scheme the class represents. This would make it easier to quickly understand the differences between the partitioning scheme classes, which would make it easier for you to present your work and it would be easier for future development/review of the classes.

File: src/main/java/org/apache/sysds/runtime/controlprogram/paramserv/FederatedPSControlThread.java [hunk 3]
```diff
@@ -58,48 +61,68 @@
...
-		if(_runtimeBalancing == Statement.PSRuntimeBalancing.SCALE_BATCH 
-			|| _runtimeBalancing == Statement.PSRuntimeBalancing.SCALE_BATCH_AND_WEIGH) {
-			throw new NotImplementedException();
+		// Calculate possible batches with batch size
+		_possibleBatchesPerLocalEpoch = (int) Math.ceil((double) dataSize / _batchSize);
+
+		// If no runtime balancing is specified, just run possible number of batches
+		// WARNING: Will get stuck on miss match
+		if(_runtimeBalancing == Statement.PSRuntimeBalancing.NONE) {
+			_numBatchesPerEpoch = _possibleBatchesPerLocalEpoch;
```
[the remark this answers is not in the document]
Comment (tobiasrieger) [reply src/main/java/org/apache/sysds/runtime/controlprogram/paramserv/FederatedPSControlThread.java:119]: Thanks for taking a look! A missmatch occurs when you specify a batch size and two workers have a different number of examples. For Example batchsize: 1 Worker 1: 20 and Worker 2: 30 examples. Without data partitioning or runtime balancing Worker 2 will wait for Worker 1 to complete batch 21 which it never does therefor getting stuck. I'll think about adding a detection feature and an exception in the next PR!

**Comment by tobiasrieger** [comment]
@sebwrede I have addressed all other proposed changes in the new commit. Thanks!

File: src/main/java/org/apache/sysds/runtime/controlprogram/paramserv/FederatedPSControlThread.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (sebwrede) [reply src/main/java/org/apache/sysds/runtime/controlprogram/paramserv/FederatedPSControlThread.java:119]: OK, that sounds like a good idea.

**Comment by sebwrede** [comment]
LGTM!

**Comment by mboehm7** [comment]
LGTM - thanks @tobiasrieger for the patch and @sebwrede for the review. I only modified some formatting issues, reduced the visibility of some members, change the string concatenation, and guarded the info logging (to avoid unnecessary string concatenation).

**Pull request closed without merging**